doc/developer/design/20210505_kafka_keys_in_sql.md

```diff
@@ -0,0 +1,165 @@
+the Key is semantically meaningful to the Kafka system -- it has effects on
+which shard the message goes to, and affects old-message deletion. We rely on
+the Key primarily for Upsert logic, ensuring deduplication of messages based on
+it.
+
+An important aspect of the Key is that it is a separate data section -- it never
+shares backing storage with the Value. This means that it is possible to have
+data in the Key that appears nowhere in the Value, putting the responsibility on
+consumers or [Kafka connectors](https://kafka-tutorials.confluent.io/connect-add-key-to-source/kafka.html) to stitch the data back together.
+
+Materialize does not support accessing the Key part of messages from our SQL
+layer, which has been mentioned as a pain point on several occasions. This is a
+design to resolve that pain point.
+
+[ddb]: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodbstreams.html#DynamoDBStreams.Client.get_records
+
+## Goals
+
+<!--
+// Enumerate the concrete goals that are in scope for the project.
+-->
+
+Allowing users to specify that a Key should be part of the dataflow, and
+accessible to all SQL processing.
+
+## Non-Goals
+
+<!--
+// Enumerate potential goals that are explicitly out of scope for the project
+// ie. what could we do or what do we want to do in the future - but are not doing now
+-->
+
+* Integrating Key descriptions more tightly into planning or index selection.
+* Any specific optimizations
+
+### Related *future* work
+
+We have two other projects that we know that we would eventually like to
+implement that may have interactions with this design:
+
+* We want to be able to specify a subset of fields from values as the keys for
+  upsert -- although current thinking is that this should probably be done as an
+  upsert function.
+* We would like to expose message metadata as columns or records to the SQL.
+
+This project does not aim to resolve any aspect of those future projects, but
+anything that conflicts with those goals in this design should definitely be
+called out.
+
+## Description
+
+<!--
+// Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+// If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+### Syntax and user-visible semantics
+
+Enhance the Kafka create source syntax like so:
+
+```sql
+CREATE SOURCE <source-name> FROM KAFKA BROKER '<>'
+<format>
+[INCLUDE KEY [AS RECORD <key-record-name>]]
+```
+
+where `<format>` is either:
+
+* An explicit key/value format:
+
+  ```sql
+  KEY FORMAT <specifier>
+  VALUE FORMAT <specifier>
+  ```
+
+* A bare format that specifies a Confluent Schema Registry (CSR) config
+
+  ```sql
+  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY ...
+  ```
+
+If `INCLUDE KEY` is specified, the source schema will be augmented to prepend
+all of the values of the key as columns in the dataflow. Format types that do
```

Comment:
Can you include an example query that would demonstrate how the view part would look?

Reply:
done